x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/ClusterDeprecationChecks.java

```diff
@@ -210,28 +212,56 @@ static DeprecationIssue checkPollIntervalTooLow(ClusterState state) {
         return null;
     }
 
-    static DeprecationIssue checkTemplatesWithMultipleTypes(ClusterState state) {
-        Set<String> templatesWithMultipleTypes = new HashSet<>();
+    static DeprecationIssue checkTemplatesWithCustomAndMultipleTypes(ClusterState state) {
+        Set<String> templatesWithMultipleTypes = new TreeSet<>();
+        Set<String> templatesWithCustomTypes = new TreeSet<>();
         state.getMetadata().getTemplates().forEach((templateCursor) -> {
             String templateName = templateCursor.key;
             ImmutableOpenMap<String, CompressedXContent> mappings = templateCursor.value.mappings();
-            if (mappings != null && mappings.size() > 1) {
-                templatesWithMultipleTypes.add(templateName);
+            if (mappings != null) {
+                if (mappings.size() > 1) {
+                    templatesWithMultipleTypes.add(templateName);
+                }
+                boolean hasCustomType = mappings.stream().anyMatch(mapping -> {
+                    String typeName = mapping.getKey();
+                    return MapperService.SINGLE_MAPPING_NAME.equals(typeName) == false;
+                });
+                if (hasCustomType) {
+                    templatesWithCustomTypes.add(templateName);
+                }
             }
         });
-        if (templatesWithMultipleTypes.isEmpty()) {
-            return null;
+        final DeprecationIssue deprecationIssue;
+        if (templatesWithMultipleTypes.isEmpty() && templatesWithCustomTypes.isEmpty()) {
+            deprecationIssue = null;
+        } else if (templatesWithMultipleTypes.isEmpty()) {
+            deprecationIssue = new DeprecationIssue(
+                DeprecationIssue.Level.CRITICAL,
+                "Custom mapping types in index templates are deprecated",
+                "https://ela.st/es-deprecation-7-custom-types",
+                "Update or remove the following index templates before upgrading to 8.0: "
+                    + templatesWithCustomTypes
+                    + ". See https://ela.st/es-deprecation-7-removal-of-types for alternatives to mapping types.",
+                false,
+                null
+            );
+        } else {
+            // There were multiple mapping types, so at least one of them had to be a custom type as well
+            Set<String> allBadTemplates = new TreeSet<>();
+            allBadTemplates.addAll(templatesWithMultipleTypes);
+            allBadTemplates.addAll(templatesWithCustomTypes);
+            deprecationIssue = new DeprecationIssue(
+                DeprecationIssue.Level.CRITICAL,
+                "Multiple mapping types and custom mapping types in index templates and indices are deprecated",
+                "https://ela.st/es-deprecation-7-multiple-types",
+                "Update or remove the following index templates before upgrading to 8.0: "
+                    + templatesWithMultipleTypes
```

Comment:
Should this be `allBadTemplates` instead?

Reply:
Good catch. I'll fix that.